lib/SIL/IR/SILValue.cpp

```diff
@@ -330,3 +320,70 @@ SILFunction *Operand::getParentFunction() const {
   auto *self = const_cast<Operand *>(this);
   return self->getUser()->getFunction();
 }
+
+bool Operand::canAcceptKind(ValueOwnershipKind kind) const {
+  auto constraint = getOwnershipConstraint();
+  if (!constraint)
+    return false;
+
+  if (constraint->satisfiesConstraint(kind))
+    return true;
+
+  // Then see if our preferred ownership constraint was not guaranteed or our
+  // use lifetime constraint was LifetimeEnding. If it wasn't, then we fail
+  // since we do not allow for implicit borrows in such situations.
+  if (kind == OwnershipKind::Owned && !isLifetimeEnding()) {
+    // Otherwise, we now know that our constraint is non lifetime ending
+    // and guaranteed and our value had owned ownership. If our user
+    // allows for implicit borrows and thus can accept owned values,
+    // return true.
+    if (auto borrowingOperand = BorrowingOperand::get(this))
+      if (borrowingOperand->canAcceptOwnedValues())
+        return true;
+  }
+
+  return false;
+}
+
+bool Operand::satisfiesConstraints() const {
+  return canAcceptKind(get().getOwnershipKind());
+}
+
+bool Operand::isLifetimeEnding() const {
+  auto constraint = getOwnershipConstraint();
+
+  // If we got back none, then our operand is for a type dependent operand. So
+  // return false.
+  if (!constraint)
+    return false;
+
+  // If our use lifetime constraint is NonLifetimeEnding, just return false.
+  if (!constraint->isLifetimeEnding())
+    return false;
+
+  // Otherwise, we may have a lifetime ending use. If we narrowed to None then
+  // we have a non lifetime ending use, otherwise we still have a lifetime
+  // ending use. If our value was not None, then obviously narrowing didn't
+  // happen.
```

Comment:
This comment talks about "narrowing", but I don't know what's being narrowed or where it's being narrowed.